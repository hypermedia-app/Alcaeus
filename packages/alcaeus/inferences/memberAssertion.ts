import type { DatasetCore, Quad } from '@rdfjs/types'
import { hydra } from '@tpluscode/rdf-ns-builders'
import { HydraEnvironment } from 'alcaeus-core'

export function * addExplicitStatementsInferredFromMemberAssertion(dataset: DatasetCore, env: HydraEnvironment): Iterable<Quad> {
  for (const collection of env.clownface({ dataset }).has([hydra.manages, hydra.memberAssertion]).toArray()) {
    const assertions = collection.out([hydra.manages, hydra.memberAssertion]).toArray()

    for (const member of collection.out(hydra.member).toArray()) {
      for (const assertion of assertions) {
        let blanks = 0

        let subject = assertion.out(hydra.subject)
        let predicate = assertion.out(hydra.property)
        let object = assertion.out(hydra.object)

        if (subject.terms.length === 0) {
          subject = member
          blanks++
        }

        if (predicate.terms.length === 0) {
          predicate = member
          blanks++
        }

        if (object.terms.length === 0) {
          object = member
          blanks++
        }

        if (blanks === 1) {
          for (const s of subject.terms) {
            for (const p of predicate.terms) {
              for (const o of object.terms) {
                yield env.quad(<any>s, <any>p, <any>o)
              }
            }
          }
        }
      }
    }
  }
}
