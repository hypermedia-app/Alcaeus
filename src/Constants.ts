export const Headers = {
    Link: 'Link',
    Location: 'Location',
    ContentType: 'Content-Type',
}

export const LinkRelations = {
    canonical: 'canonical',
}
