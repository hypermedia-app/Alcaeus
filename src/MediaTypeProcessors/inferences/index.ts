export { inferTypesFromPropertyRanges } from './propertyTypes'
export { addExplicitStatementsInferredFromManagesBlock } from './managesBlock'
export { inheritFromSuperclasses } from './subclassOfInherit'
