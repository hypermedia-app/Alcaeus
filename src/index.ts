import {HydraResource as ResourceCtor} from "./Resources";
import {ResourceFactory as ResourceFactoryCtor} from './ResourceFactory';
import {Heracles} from './heracles';

export let ResourceFactory = ResourceFactoryCtor;
export let Resource = ResourceCtor;
export let Hydra = new Heracles();