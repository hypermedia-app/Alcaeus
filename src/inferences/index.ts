export { inferTypesFromPropertyRanges } from './propertyTypes.js'
export { addExplicitStatementsInferredFromMemberAssertion } from './memberAssertion.js'
