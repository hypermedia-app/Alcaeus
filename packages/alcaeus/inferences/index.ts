import { inferTypesFromPropertyRanges } from './propertyTypes.js'
import { addExplicitStatementsInferredFromMemberAssertion } from './memberAssertion.js'

export default [
  inferTypesFromPropertyRanges,
  addExplicitStatementsInferredFromMemberAssertion,
]
