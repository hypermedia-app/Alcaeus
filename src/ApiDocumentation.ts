import {Core, JsonLd} from './Constants';
import {Schema, rdfs, owl} from './Vocabs';
import {Resource} from './Resources';
import {nonenumerable} from "core-decorators";
import {
    IApiDocumentation, IHeracles, IClass, ISupportedOperation, ISupportedProperty, IHydraResource,
    IDocumentedResource, IStatusCodeDescription, IRdfProperty, IResource
} from './interfaces';

const heraclesWeakMap = new WeakMap<IResource, IHeracles>();

export class ApiDocumentation extends Resource implements IApiDocumentation {

    constructor(heracles:IHeracles, apiDoc:any) {
        super(apiDoc);

        heraclesWeakMap.set(this, heracles);
    }

    get classes():Array<IClass> {
        if (Array.isArray(this[Core.Vocab.supportedClass])) {
            return this[Core.Vocab.supportedClass];
        }

        return [ this[Core.Vocab.supportedClass] ];
    }

    @nonenumerable
    get _heracles():IHeracles {
        return heraclesWeakMap.get(this);
    }

    getOperations(classUri:string):Array<ISupportedOperation>;

    getOperations(classUri:string, predicateUri?:string):Array<ISupportedOperation> {
        let clas = this.getClass(classUri);
        if (!clas) {
            return [];
        }
        
        if(!predicateUri) {
            return clas.supportedOperations;
        }

        let supportedProperty = clas.supportedProperties.find((prop: ISupportedProperty) => {
            return prop.property && prop.property.id === predicateUri;
        });
        if(!supportedProperty) {
            return [];
        }

        return supportedProperty.property.supportedOperations;
    }

    getProperties(classUri:string):Array<ISupportedProperty> {
        let clas = this.getClass(classUri);
        if (!clas) {
            return [];
        }
        return clas.supportedProperties;
    }

    getClass(classId):IClass {
        return this.classes.find(clas => clas[JsonLd.Id] === classId) || null;
    }

    getEntrypoint():Promise<IHydraResource> {
        return this._heracles.loadResource(this[Core.Vocab.entrypoint][JsonLd.Id]);
    }
}

export class DocumentedResource extends Resource implements IDocumentedResource {
    constructor(hydraResource:any) {
        super(hydraResource);
    }

    get description():string {
        return this[Core.Vocab.description] ||
            this[rdfs.comment] ||
            this[Schema.description]
    }

    get title():string {
        return this[Core.Vocab.title] ||
            this[rdfs.label] ||
            this[Schema.title];
    }
}

export class SupportedOperation extends DocumentedResource implements ISupportedOperation {

    constructor(hydraOperation:any, heracles:IHeracles) {
        super(hydraOperation);
        heraclesWeakMap.set(this, heracles);
    }

    get method():string {
        return this[Core.Vocab.method];
    }

    get expects():IClass {
        return this[Core.Vocab.expects];
    }

    get returns():IClass {
        return this[Core.Vocab.returns];
    }

    get requiresInput():boolean {
        const method = this.method || '';
        const methodExpectsBody = method.toUpperCase() !== 'GET' && this.method.toUpperCase() !== 'DELETE';

        const operationExpectsBody = !!this.expects && this.expects.id !== owl.Nothing;

        return methodExpectsBody || operationExpectsBody;
    }
}

export class SupportedProperty extends DocumentedResource implements ISupportedProperty {

    constructor(hydraSupportedProperty:any) {
        super(hydraSupportedProperty);
    }

    get readable() {
        if (typeof this[Core.Vocab.readable] === 'boolean') {
            return this[Core.Vocab.readable];
        }

        return true;
    }

    get writable() {
        if (typeof this[Core.Vocab.writable] === 'boolean') {
            return this[Core.Vocab.writable];
        }

        return true;
    }

    get required() {
        if (typeof this[Core.Vocab.required] === 'boolean') {
            return this[Core.Vocab.required];
        }

        return false;
    }

    get property() {
        return this[Core.Vocab.property];
    }
}

export class Class extends DocumentedResource implements IClass {

    constructor(hydraClass:Object) {
        super(hydraClass);
    }

    get supportedOperations():Array<ISupportedOperation> {
        let operations = this[Core.Vocab.supportedOperation];
        
        if(typeof operations ==='undefined' || operations === null) {
            return [];
        }
        
        if(Array.isArray(operations)) {
            return operations;
        }

        return [ operations ];
    }

    get supportedProperties():Array<ISupportedProperty> {
        let properties = this[Core.Vocab.supportedProperty];
        
        if(typeof properties === 'undefined' || properties === null ) {
            return [];
        }
        
        if(Array.isArray(properties)) {
            return properties;
        }

        return [ properties ];
    }
}

export class StatusCodeDescription extends Resource implements IStatusCodeDescription {
    
    get code():number {
        return this[Core.Vocab.code];
    }
    
    get description():string {
        return this[Core.Vocab.description] || '';
    }
}

export class RdfProperty extends DocumentedResource implements IRdfProperty {
    get range():IClass {
        return this[rdfs.range];
    }

    get domain():IClass {
        return this[rdfs.domain];
    }

    get supportedOperations():Array<ISupportedOperation> {
        let value = this[Core.Vocab.supportedOperation];

        if(typeof value === 'undefined'){
            return [];
        }

        if(Array.isArray(value) === false) {
            return [ value ];
        }

        return this[Core.Vocab.supportedOperation];
    }
}