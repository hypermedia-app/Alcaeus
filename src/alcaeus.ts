import { RdfResource, ResourceFactory } from '@tpluscode/rdfine'
import $rdf from 'rdf-ext'
import DatasetExt from 'rdf-ext/lib/Dataset'
import { NamedNode } from 'rdf-js'
import TripleToQuadTransform from 'rdf-transform-triple-to-quad'
import * as FetchUtil from './FetchUtil'
import { merge } from './helpers/MergeHeaders'
import { HydraResponse, create } from './HydraResponse'
import { MediaTypeProcessor } from './MediaTypeProcessors/RdfProcessor'
import { ApiDocumentation, HydraResource } from './Resources'
import { Operation } from './Resources/Operation'
import { ResponseWrapper } from './ResponseWrapper'
import { RootSelector } from './RootSelectors'

export interface HydraClient {
    rootSelectors: RootSelector[];
    mediaTypeProcessors: { [name: string]: MediaTypeProcessor };
    loadResource(uri: string, headers?: HeadersInit): Promise<HydraResponse>;
    loadDocumentation (uri: string, headers: HeadersInit): void;
    invokeOperation(operation: Operation, uri: string, body?: BodyInit, headers?: string | HeadersInit): Promise<HydraResponse>;
    defaultHeaders: HeadersInit | (() => HeadersInit);
    dataset: DatasetExt;
    factory: ResourceFactory;
    apiDocumentations: Promise<ApiDocumentation[]>;
}

const addOrReplaceGraph = async (
    alcaeus: HydraClient,
    response: ResponseWrapper,
    uri: string): Promise<void> => {
    const suitableProcessor = Object.values(alcaeus.mediaTypeProcessors)
        .find((processor) => processor.canProcess(response.mediaType))

    if (!suitableProcessor) {
        console.warn(`No processor found for media type ${response.mediaType}`)
        return
    }

    const graph = $rdf.namedNode(uri)
    const parsedTriples = await suitableProcessor.process(uri, response)
    await alcaeus.dataset
        .removeMatches(undefined, undefined, undefined, graph)
        .import(parsedTriples.pipe(new TripleToQuadTransform(graph)))
}

export class Alcaeus<R extends HydraResource> implements HydraClient {
    public rootSelectors: RootSelector[];

    public mediaTypeProcessors: { [name: string]: MediaTypeProcessor };

    public defaultHeaders: HeadersInit | (() => HeadersInit) = {}

    public readonly dataset: DatasetExt = $rdf.dataset()

    public readonly factory: ResourceFactory<R>

    private readonly __documentationPromises: Map<string, Promise<HydraResponse>> = new Map()

    public constructor (
        rootSelectors: RootSelector[],
        mediaTypeProcessors: { [name: string]: MediaTypeProcessor },
        factory: ResourceFactory<R>
    ) {
        this.rootSelectors = rootSelectors
        this.mediaTypeProcessors = mediaTypeProcessors
        this.factory = factory
    }

    public get apiDocumentations () {
        return Promise.all(this.__documentationPromises.values())
            .then(responses => responses.map<ApiDocumentation | RdfResource | null>(r => r.root))
            .then(docs => {
                return docs.reduce((notNull, doc) => {
                    if (doc && 'loadEntrypoint' in doc) {
                        notNull.push(doc)
                    }

                    return notNull
                },
                [] as ApiDocumentation[])
            })
    }

    public async loadResource (id: string | NamedNode, headers: HeadersInit = {}): Promise<HydraResponse> {
        const uri = typeof id === 'string' ? id : id.value

        const response = await FetchUtil.fetchResource(uri, this.__mergeHeaders(new Headers(headers)))
        await addOrReplaceGraph(this, response, uri)
        this.__getApiDocumentation(response, headers)

        return create(uri, response, this)
    }

    public loadDocumentation (uri: string, headers: HeadersInit = {}) {
        const request = FetchUtil.fetchResource(uri, this.__mergeHeaders(new Headers(headers)))
            .then(async response => {
                await addOrReplaceGraph(this, response, uri)
                return response
            })
            .then(response => create(uri, response, this))

        this.__documentationPromises.set(uri, request)
    }

    public async invokeOperation (operation: Pick<Operation, 'method'>, uri: string, body?: BodyInit, headers: HeadersInit = {}): Promise<HydraResponse> {
        const mergedHeaders = this.__mergeHeaders(new Headers(headers))

        const response = await FetchUtil.invokeOperation(operation.method, uri, body, mergedHeaders)
        this.__getApiDocumentation(response, headers)

        return create(uri, response, this)
    }

    private __getApiDocumentation (response: ResponseWrapper, headers: HeadersInit) {
        if (!response.apiDocumentationLink) {
            console.warn(`Resource ${response.requestedUri} does not expose API Documentation link`)
            return
        }

        this.loadDocumentation(response.apiDocumentationLink, headers)
    }

    private __mergeHeaders (headers: Headers): Headers {
        const defaultHeaders = typeof this.defaultHeaders === 'function' ? this.defaultHeaders() : this.defaultHeaders

        return merge(new Headers(defaultHeaders), headers)
    }
}
