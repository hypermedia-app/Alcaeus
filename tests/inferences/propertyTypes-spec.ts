import { inferTypesFromPropertyRanges } from '../../src/inferences'
import * as specGraphs from './propertyTypes-spec-graphs'

describe('property types inference', () => {
    it('adds rdf:type assertions for known properties', async () => {
        // given
        const dataset = await specGraphs.managesWithType()

        // when
        dataset.addAll([...inferTypesFromPropertyRanges(dataset)])

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
    })
})
