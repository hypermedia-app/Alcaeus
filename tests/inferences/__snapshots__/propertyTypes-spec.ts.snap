// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`property types inference adds rdf:type assertions for known properties 1`] = `
"<http://example.com/API> <http://www.w3.org/ns/hydra/core#supportedClass> <http://example.com/Class> .
<http://example.com/Class> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://example.com/Class> <http://www.w3.org/ns/hydra/core#supportedOperation> <http://example.com/Operation> .
<http://example.com/Class> <http://www.w3.org/ns/hydra/core#supportedProperty> <http://example.com/Property> .
<http://example.com/ExpectedClass> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://example.com/IriTemplateMapping> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#IriTemplateMapping> .
<http://example.com/Operation> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#SupportedOperation> .
<http://example.com/Operation> <http://www.w3.org/ns/hydra/core#expects> <http://example.com/ExpectedClass> .
<http://example.com/Operation> <http://www.w3.org/ns/hydra/core#returns> <http://example.com/ReturnedClass> .
<http://example.com/Operation> <http://www.w3.org/ns/hydra/core#statusCodes> _:c14n0 .
<http://example.com/OperationInstance> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Operation> .
<http://example.com/Property> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#SupportedProperty> .
<http://example.com/Property> <http://www.w3.org/ns/hydra/core#property> <http://xmlns.com/foaf/0.1/knows> .
<http://example.com/Resource> <http://www.w3.org/ns/hydra/core#mapping> <http://example.com/IriTemplateMapping> .
<http://example.com/Resource> <http://www.w3.org/ns/hydra/core#operation> <http://example.com/OperationInstance> .
<http://example.com/ReturnedClass> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://xmlns.com/foaf/0.1/knows> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/1999/02/22-rdf-syntax-ns#Property> .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#StatusCodeDescription> .
"
`;
